### PR TITLE
[GUVNOR-2279] Deploy artifact with encrypted password.

### PR DESCRIPTION
**Pull request by afcollins**
When deploying an artifact to a remote repository and the server
password is encrypted using master password, reuse the Authentication
that is built within the RepositorySystemSession.

Depends on: https://github.com/droolsjbpm/drools/pull/538

### CODE REVIEW

File: guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 1]
```diff
@@ -509,11 +513,9 @@ private RemoteRepository getRemoteRepoFromDeployment( DeploymentRepository repo
         Server server = settings.getServer( repo.getId() );
 
         if ( server != null ) {
-            remoteRepoBuilder.setAuthentication(
-                    new AuthenticationBuilder()
-                            .addUsername( server.getUsername() )
-                            .addPassword( server.getPassword() )
-                            .build() );
+            Authentication authentication = embedder.getMavenSession().getRepositorySession()
+                    .getAuthenticationSelector().getAuthentication( remoteRepoBuilder.build() );
```
Comment (manstis) [inline guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: Hi, I'm a little concerned that this PR may not be using org.eclipse.aether correctly; the [Java Docs](http://download.eclipse.org/aether/aether-core/1.0.0/apidocs/org/eclipse/aether/DefaultRepositorySystemSession.html#getAuthenticationSelector()) state "Note that this selector is not used for remote repositories..." whereas we seem to be specifically using it to configure remotes' Authentication. Am I missing something?

**Comment by manstis** [comment]
We'd really, really, need some form of test in the PR. Is there nothing that can be done re: Aether?

**Comment by afcollins** [comment]
@manstis I attached a maven project with a test to the JIRA.  It contains a duplicate Aether class that allows me to mock the fields.  It was how I was testing the fix, but did not think it would fit to go as a test in this module as the Aether was causing other test failures.  I can create the project as new module of guvnor-m2repo-editor, but wanted to keep out of PR.  Please let me know if I should add to PR.

https://issues.jboss.org/secure/attachment/12394505/guvnor-m2repo-editor-test.zip

File: guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java (hunk 1 above)
Comment (afcollins) [inline guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: Hmm, the aether javadoc is the orginal from 2011.  May have been that way for Eclipse, but Maven uses AuthenticationSelector is same way while building the project (regardless of mojos executed) at [AuthenticationSelector (line 246)](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=blob;f=maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java;h=93317105eb196040fc310a28a0e4c00c2bd36ca8;hb=HEAD).  This was part of inspiration to use AuthenticationSelector, as the Authentication itself is encoded in memory, and reading the settings-security.xml to manually decrypt passwords would duplicate work already accomplished by DefaultMaven (invoked in MavenEmbedder constructor).

Comment (manstis) [inline guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: @afcollins Thanks for the explanation and link. When @mariofusco has confirmed he's happy with the other (Drools) PR I can request this be merged into master too - I might fork your fork and try integrating your test into the existing module (IDK what I can do that you have not tried, but I'd prefer the test in the same module as the change - and using the existing Aether.. if possible).

Comment (manstis) [inline guvnor-m2repo-editor/guvnor-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: Drools PR has been merged. I'll look at this tomorrow (Thursday GMT).
Thanks for the PR.. it'll be great to cover your use case.

**Comment by manstis** [comment]
@afcollins I grabbed your branch locally and ran it. The test fails:

```
testDeployArtifactWitdeployArtifacthDistributionManagement(org.guvnor.m2repo.backend.server.GuvnorM2RepositoryTest)  Time elapsed: 1.049 sec  <<< FAILURE!
java.lang.AssertionError: Plaintext pw (repopw) length expected. expected:<6> but was:<46>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.guvnor.m2repo.backend.server.GuvnorM2RepositoryTest$3.matches(GuvnorM2RepositoryTest.java:149)
	at org.mockito.internal.matchers.LocalizedMatcher.matches(LocalizedMatcher.java:30)
	at org.mockito.internal.invocation.ArgumentsComparator.argumentsMatch(ArgumentsComparator.java:26)
	at org.mockito.internal.invocation.ArgumentsComparator.argumentsMatch(ArgumentsComparator.java:18)
	at org.mockito.internal.invocation.InvocationMatcher.matches(InvocationMatcher.java:58)
	at org.mockito.internal.invocation.InvocationsFinder$RemoveNotMatching.isOut(InvocationsFinder.java:145)
	at org.mockito.internal.invocation.InvocationsFinder$RemoveNotMatching.isOut(InvocationsFinder.java:137)
	at org.mockito.internal.util.collections.ListUtil.filter(ListUtil.java:16)
	at org.mockito.internal.invocation.InvocationsFinder.findInvocations(InvocationsFinder.java:20)
	at org.mockito.internal.verification.checkers.MissingInvocationChecker.check(MissingInvocationChecker.java:32)
	at org.mockito.internal.verification.Times.verify(Times.java:36)
	at org.mockito.internal.verification.MockAwareVerificationMode.verify(MockAwareVerificationMode.java:21)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:72)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:38)
	at org.mockito.internal.creation.MethodInterceptorFilter.intercept(MethodInterceptorFilter.java:51)
	at org.eclipse.aether.RepositorySystem$$EnhancerByMockitoWithCGLIB$$aca7867.deploy(<generated>)
	at org.guvnor.m2repo.backend.server.GuvnorM2RepositoryTest.testDeployArtifactWitdeployArtifacthDistributionManagement(GuvnorM2RepositoryTest.java:119)

org.guvnor.m2repo.backend.server.GuvnorM2RepositoryTest  Time elapsed: 1.05 sec  <<< ERROR!
java.lang.NullPointerException: null
	at java.util.Hashtable.put(Hashtable.java:514)
	at java.util.Properties.setProperty(Properties.java:161)
	at java.lang.System.setProperty(System.java:787)
	at org.guvnor.m2repo.backend.server.GuvnorM2RepositoryTest.tearDownClass(GuvnorM2RepositoryTest.java:67)
```

**Comment by manstis** [comment]
@afcollins hehe.. PEBKAC...

**Comment by afcollins** [comment]
hehe, just double checked latest kie-ci snapshots (build 332), and the
commit is in that build. test passes for me with maven build, but failing
in IDE. Taking a look.

**Comment by afcollins** [comment]
@manstis test case should not fail now. I was seeing failure due to split intellij project directory, but it should pass now regardless of execution environment.

**Comment by manstis** [comment]
@afcollins Thanks. I had a fiddle and managed to mock use of the *real* ```Aether```. See https://github.com/manstis/guvnor/commit/72bd4e0a69bdb5651a9aafbc740aeddcca82423e. That said the test is failing (the password is 46 characters and not 6).. Have you see this type of failure before? Am I doing something daft? I'd prefer to use the existing ```Aether``` if possible. Thanks, Mike

**Comment by afcollins** [comment]
@manstis Yes, that is the test of whether master password is being used to decrypt server password.  I will take a look at your commit, and integrate into backend tests.

**Comment by afcollins** [comment]
@manstis when I debug your commit, I see a ClassCastException thrown from javax.crypto.Cipher.chooseProvider().  From what I can find, using both @RunsWith(PowerMockRunner.class) and @PrepareForTest(Aether.class) will yield the ClassCastException, though only using one or the other will not.  I have added three tests at afcollins@578f8223914abde691dd0614b807ca3dfc69a3b5 to demonstrate that only when both annotations are used, the ClassCastException occurs.

I am unable to debug the actual ClassCastException itself, but I suspect that powermock's bytecode manipulation is interfering with PBECipher.General bytecode.

As long as this exception is thrown, the server password will not be decrypted as expected.

**Comment by manstis** [comment]
@afcollins 

Hi, thanks for sticking with me on this! 

I had a look and found (thanks Google!) that the issue can be avoided with ```@PowerMockIgnore({ "javax.crypto.*" })```. I've updated my [commit](https://github.com/manstis/guvnor/commit/71908fe4eebc5a62be748e583859dcace5f94d5c). Could you please revise your PR to use the mocked ```Aether``` and move the test into the existing ```guvnor-m2repo-editor-backend``` module's test package? I also noticed a few cosmetic formatting issues in your PR. Could you please review your formatting too (normally I'd be happy to fix formatting issues, but I'm unable commit to ```droolsjbpm/guvnor``` either now we've moved to an internal PR-review process too).

Happy to agree to commit after this.

Thanks,

Mike

**Comment by afcollins** [comment]
@manstis I will be happy to incorporate the test into *backend module.  Re: formatting: I followed all instructions at https://github.com/droolsjbpm/droolsjbpm-build-bootstrap#developing-with-eclipse (same for intellij) and I committed the PR with the java formatting I imported from these instructions.  From your last commit I have reverse-engineered your formatting standards, but is there a formatting template I should be using published elsewhere?  Otherwise, I am not guaranteed my reverse-engineered settings will be 100% accurate.

**Comment by manstis** [comment]
@afcollins Hi,

Code Formatters can be found [here](https://github.com/droolsjbpm/droolsjbpm-build-bootstrap/tree/master/ide-configuration/eclipse-configuration/code-style)

The thing that jumped out at me was the split ASL license [here](https://github.com/droolsjbpm/guvnor/pull/174/files#diff-7ae7086cfd233830feafe41c811d1700R1)

Cheers,

Mike

**Comment by afcollins** [comment]
@manstis Yes, those are the formatters I imported initially.  When I import all three formatters, and Format Code and Organize Imports in eclipse on your commit, there are many formatting changes (alignment, spaces, etc.): https://github.com/afcollins/guvnor/commit/7170069a62bfb3a7c239e8aff3a455729fa9c9e2  most noticeably, your file has spaces between parens and arguments in a method invocation.  Would you please update the formatters with your current setup?

Also, was thinking I would rebase my topic branch to master, and submit via a new PR once everything is final.  Is this OK, or should I push without syncing up with master?

**Comment by manstis** [comment]
I use IDEA... sorry cannot update the Eclipse ones. Provided you're using
the formatter and you correct the placement of the license I think we'll be
fine. A new PR is great too. Thanks.

**Comment by afcollins** [comment]
Same happens from IDEA. Could you update the idea formatter in
build-bootstrap project?

**Comment by afcollins** [comment]
Submitted squashed PR https://github.com/droolsjbpm/guvnor/pull/187.

**Comment by manstis** [comment]
@porcelli @csadilek Please close this PR. It has been superseded by https://github.com/droolsjbpm/guvnor/pull/187

**Comment by psiroky** [comment]
Closing as requested.

**Pull request closed without merging**